metrics/exporter.go

```diff
@@ -82,17 +105,48 @@ func UpdateExporter(ops ExporterOptions, logger *zap.SugaredLogger) error {
 	metricsMux.Lock()
 	defer metricsMux.Unlock()
 
-	newConfig, err := createMetricsConfig(ops, logger)
-	if err != nil {
-		if curMetricsExporter == nil {
-			// Fail the process if there doesn't exist an exporter.
-			logger.Errorw("Failed to get a valid metrics config", zap.Error(err))
-		} else {
-			logger.Errorw("Failed to get a valid metrics config; Skip updating the metrics exporter", zap.Error(err))
+	if ops.StackdriverConfigMap != nil {
+		sc, sErr := sdconfig.NewStackdriverConfigFromMap(ops.StackdriverConfigMap)
+		if sErr != nil {
+			logger.Warnw("Failed to get a valid Stackdriver config", zap.Error(sErr))
 		}
-		return err
+
+		// Always store updates from Stackdriver config map, even if exporter
+		// itself cannot be updated yet.
+		curStackdriverConfig = sc
 	}
 
+	// The stackdriver config alone is not sufficient to create a metrics exporter.
+	// Wait for the first update to ExporterOptions.ConfigMap to update the exporter.
+	if ops.ConfigMap == nil && curMetricsConfig == nil {
+		return nil
+	}
+
+	var newConfig *metricsConfig
+	if ops.ConfigMap != nil {
+		// If there's an updated config map, build a new config from scratch.
+		nc, err := createMetricsConfig(ops, logger)
+		if err != nil {
+			if curMetricsExporter == nil {
+				// Fail the process if there doesn't exist an exporter.
+				logger.Errorw("Failed to get a valid metrics config", zap.Error(err))
+			} else {
+				logger.Errorw("Failed to get a valid metrics config; Skip updating the metrics exporter", zap.Error(err))
+			}
+			return err
+		}
+
+		newConfig = nc
+	} else {
+		// Otherwise, build off copy of the current config.
+		// It is always true that curMetricsConfig is non-nil to reach here.
+		newConfig = new(metricsConfig)
+		*newConfig = *curMetricsConfig
```

Comment:
What's the reason for this `if-else` statement. Did you see some issue when `ops.ConfigMap` is null?

Reply:
The new Stackdriver configuration values alone are not enough to create a metrics exporter, so they should always be added on to an existing metricsConfig.

In the case where just the Stackdriver configuration values are changed (someone edits "config-stackdriver" but not "config-observability"), ops.ConfigMap will be null, but we want to overwrite the existing metrics config instead of creating a brand new one with default values for the non-Stackdriver configuration.

Reply:
How about 
```
newConfig := curMetricsConfig
if ops.ConfigMap != nil {
  // If there's an updated config map, build a new config from scratch.
  ...
}
```

to drop the else statement.